core/model-api/src/main/java/org/eclipse/rdf4j/model/base/CoreDatatype.java

```diff
@@ -0,0 +1,282 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Eclipse RDF4J contributors.
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Distribution License v1.0
+ *  which accompanies this distribution, and is available at
+ *  http://www.eclipse.org/org/documents/edl-v10.php.
+ ******************************************************************************/
+
+package org.eclipse.rdf4j.model.base;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Optional;
+
+import javax.xml.datatype.Duration;
+import javax.xml.datatype.XMLGregorianCalendar;
+
+import org.eclipse.rdf4j.model.IRI;
+
+public enum CoreDatatype {
```

Comment:
Can we make `CoreDatatype` implement a basic interface and then use that interface consistently? That way, we can later on add additional enum classes with other datatype vocabularies, if we want.

Also, perhaps we should call this enum "BuiltInDatatype" as they correspond to the list of XML Schema built-in datatypes defined here: https://www.w3.org/TR/xmlschema11-2/#built-in-datatypes  (plus the one extra odd case of langString).

Reply:
I had initially used an interface and made CoreDatatypes implement that interface. The reason I dropped that approach was that I got worried that if we use `==` then we would get into trouble if someone implements their own XSD datatypes and expects it to function with our code.

I'm open for other names, BuiltInDatatype would work well too. We aren't limited to XSD datatypes though, any datatype that we use directly in our code would make sense to add.

Reply:
Fair enough. l'm fine with sticking with CoreDatatype (and no interface) then, at least for this iteration.

Reply:
https://github.com/eclipse/rdf4j/issues/2643

An interface would definitely make it easier to support units of measure.

Reply:
If you want to go for an interface after all: we can stick a warning on the interface documentation to indicate restrictions in third parties extending it. Or even choose to mark it `@InternalUseOnly` for now (most normal usage shouldn't _directly_ use it, anyway). If we do go that route, I'd vote for calling this enum `BuiltInDatatype` (and document that it reflects the XML Schema built-ins), and perhaps use `CoreDatatype` for the more general interface.